### ISSUE TITLE
MySQL Database support

### ISSUE THREAD

**Issue by Sanjay-B** [nitely/Spirit#169]
Does this have MySQL Database support? If not, please consider it. 

Thanks, Sanjay.

**Comment by nitely** [comment]
From the readme:

Compatibility
* ...
* PostgreSQL (recommended), MySQL, Oracle Database and SQLite

**Comment by Sanjay-B** [comment]
My bad, thanks! :)

**Comment by nitely** [comment]
No problem :+1:

**Comment by Sanjay-B** [comment]
I ran into a problem I think you should see. 

Traceback (most recent call last):
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\backends\mysql\base.py", line 25, in <module>
    import MySQLdb as Database
ImportError: No module named 'MySQLdb'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\utils\autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\core\management\commands\runserver.py", line 113, in inner_run
    autoreload.raise_last_exception()
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\utils\autoreload.py", line 249, in raise_last_exception
    six.reraise(*_exception)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\utils\six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\utils\autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\__init__.py", line 27, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\apps\registry.py", line 108, in populate
    app_config.import_models(all_models)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\apps\config.py", line 199, in import_models
    self.models_module = import_module(models_module_name)
  File "C:\Program Files (x86)\Python35-32\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 665, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\contrib\auth\models.py", line 4, in <module>
    from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\contrib\auth\base_user.py", line 52, in <module>
    class AbstractBaseUser(models.Model):
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\models\base.py", line 119, in __new__
    new_class.add_to_class('_meta', Options(meta, app_label))
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\models\base.py", line 316, in add_to_class
    value.contribute_to_class(cls, name)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\models\options.py", line 214, in contribute_to_class
    self.db_table = truncate_name(self.db_table, connection.ops.max_name_length())
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\__init__.py", line 33, in __getattr__
    return getattr(connections[DEFAULT_DB_ALIAS], item)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\utils.py", line 211, in __getitem__
    backend = load_backend(db['ENGINE'])
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\utils.py", line 115, in load_backend
    return import_module('%s.base' % backend_name)
  File "C:\Program Files (x86)\Python35-32\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "C:\Program Files (x86)\Python35-32\lib\site-packages\django\db\backends\mysql\base.py", line 28, in <module>
    raise ImproperlyConfigured("Error loading MySQLdb module: %s" % e)
django.core.exceptions.ImproperlyConfigured: Error loading MySQLdb module: No module named 'MySQLdb'

So basically, for python 3.x users, you cannot use MySQL since that module is only compatible with Python 2.x. Not sure what to do from here. Any ideas? Its pretty much "All MySQL" or "ditch project".

**Comment by nitely** [comment]
Checking the django docs[0] there is `mysqlclient` which support py3.

[0] https://docs.djangoproject.com/en/1.10/ref/databases/#mysql-db-api-drivers
[1] https://pypi.python.org/pypi/mysqlclient

**Comment by nitely** [comment]
There is also `mysql-connector-python` which seems to be the official mysql connector, so you may want to try that one as well.

[0] https://pypi.python.org/pypi/mysql-connector-python/2.0.4

**Comment by Sanjay-B** [comment]
Ah thanks, that seemed to work. Another thing while im here, I cant seem to log into the django admin via "localhost:5000/admin". Sorry to ask so many questions. Im new to the system, not python. :P

**Comment by nitely** [comment]
That's ok. Did you created an admin user? gotta run `$ python manage.py createsuperuser` [0] or one of the alternative commands (`django-admin`, etc)

[0] https://docs.djangoproject.com/en/1.10/ref/django-admin/#usage

**Comment by Sanjay-B** [comment]
Oh wow nice! By chance, if I wanted to test the example out with 50 or so people, would it be able to work as it should? (Got the MySQL part working). Again, thanks so much. :)

**Comment by nitely** [comment]
Yes. Even on the crappiest hardware it will support 50rps.

Although having 50 people doesn't necessarily mean having 50rps, it means there is a maximum of 50rps, but it should usually be much much less than that.

**Comment by Sanjay-B** [comment]
Very nice. I changed the name from "Spirit" to my company name through the individual html files but when I run the server, it still shows up as "Spirit". Is there some setting overriding the default name or no?

Starting to feel bad asking this many questions, haha.

**Comment by nitely** [comment]
The core app contains the header template, but there is an easier way to change the title: go to spirit's admin panel, then to "basic settings" and change it from there `/st/admin/config/`

I'm unsure of what you did. I'm guessing you changed the mocks within the static folder? I made those while designing the UI but they are not used by spirit at all. The templates are spread over the many apps (threads, comments, etc).

> Starting to feel bad asking this many questions, haha.

No worries. Ask away.

**Comment by Sanjay-B** [comment]
Wow, that's neat! Is there anyway to distinct an admin/mod from a normal user in a thread or topic? ie, has a moderator/admin tag.

**Comment by nitely** [comment]
They have a subtle highlight on their user name. Blue for moderators and yellow for admins.

**Comment by Sanjay-B** [comment]
Personally, I think it would be better to have "[Admin]" or "[Mod]" next to each name that applies and have it like that when the user clicks on the target users profile. Subtle highlighting is elegant but its not as distinctive.

**Comment by Sanjay-B** [comment]
C:\Users\sanja\Desktop\Spirit\example>python manage.py runserver --settings=project.settings.prod_local
Performing system checks...

System check identified no issues (0 silenced).

You have 55 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): admin, auth, contenttypes, djconfig, sessions, spirit_category, spirit_comment, spirit_comment_bookmark, spirit_comment_flag, spirit_comment_history, spirit_comment_like, spirit_comment_poll, spirit_topic, spirit_topic_favorite, spirit_topic_notification, spirit_topic_poll, spirit_topic_private, spirit_topic_unread, spirit_user.
Run 'python manage.py migrate' to apply them.
March 02, 2017 - 20:24:31
Django version 1.10.5, using settings 'project.settings.prod_local'
Starting development server at http://127.0.0.1:8000/
Quit the server with CTRL-BREAK.
^C
C:\Users\sanja\Desktop\Spirit\example>python manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, djconfig, sessions, spirit_category, spirit_comment, spirit_comment_bookmark, spirit_comment_flag, spirit_comment_history, spirit_comment_like, spirit_comment_poll, spirit_topic, spirit_topic_favorite, spirit_topic_notification, spirit_topic_poll, spirit_topic_private, spirit_topic_unread, spirit_user
Running migrations:
  No migrations to apply.

Even when I create a prod_local and set up my own database + a host, it defaults to localhost. How would one get it to run on the address besides localhost?

**Comment by nitely** [comment]
Well if you are running: `python manage.py runserver --settings=project.settings.prod_local` then you have to add that option everywhere: `python manage.py migrate --settings=project.settings.prod_local`

> Even when I create a prod_local and set up my own database + a host, it defaults to localhost. How would one get it to run on the address besides localhost?

Are you running it locally or in a web host? if you are running locally and want to add an alias to localhost (ie. `myforum.dev`) you should edit your host file. If you are running in a web host then you should use a proper web server like gunicorn or uwsgi and likely a reverse proxy like nginx... or host it on a PaaS like heroku that does everything for you.

**Comment by Sanjay-B** [comment]
I actually host my web server off of HelioHost. Im not entirely sure on how to run it via cpanel.

**Comment by nitely** [comment]
http://www.heliohost.org/classic/features/languages/python

> Django, on the other hand, is probably the hardest scripting feature to configure on HelioHost. If you have not yet created a Django application, we suggest that you develop at home and upload to HelioHost after you are finished. This is because the use of the command-line utility django-admin.py is common during the development process, and HelioHost does not allow command-line access to its users. Development can be done without this utility, however. If it is your wish to develop on-site you will need to manually create a blank `__init__.py` file, a settings.py file, and a urls.py file. To learn more about these files try the learning links provided above.

lulz

Seems like you won't be able to create migrations either. You basically have to import your whole database from your local machine. Then you have to upload all requires python packages as well. Kinda crappy. May be you should consider other alternatives like pythonanywhere.com and openshift.com. But I can't really help you much here, how you host Spirit is out of the scope of this project.

**Comment by Sanjay-B** [comment]
I tried openshift but that was a nightmare. I was wondering if I could connect via SSH, install python, make the version of Spirit I have as a repo and just git push it. Then, I could run the command and see what happens. Though, would "allowed hosts" be the domain name or just local host in this scenario?

**Comment by Sanjay-B** [comment]
Well, nvm that idea. No command line access. Hmm, any ideas? Lel

**Comment by nitely** [comment]
HelioHost does not allow shell access, so you won't be able to run commands. Check the link I posted above. It seems you have to upload everything through FTP and import the database from a `.sql` file.

> I tried openshift but that was a nightmare.

There is a one click deployment for heroku[0]. However the free plan has some serious limitations.

> Though, would "allowed hosts" be the domain name or just local host in this scenario?

The domain.

[0] https://github.com/nitely/spirit-heroku

**Comment by Sanjay-B** [comment]
So Heroku + HelioHost should be fine right? So, I use the domain for HelioHost and run the application from Heroku, correct?

**Comment by nitely** [comment]
They gave you a domain? or a sub domain (ie, `mysite.heliohost.org`)

If they somehow gave you a domain and you can change its DNS, then sure. Be aware heroku free plan is crap. They tear down the web app after 30 minutes of being idle and it takes a few seconds to spin back up on the first request. Also, they tear down the web app for at least 6hs a day if the account is not verified (no credit card attach to it).

**Comment by Sanjay-B** [comment]
So what are my options to running this without heroku? No, I got a sub-domain. Sadly.

**Comment by nitely** [comment]
> No, I got a sub-domain. Sadly.

Then no...

Heroku gives you a subdomain as well. You can try pythonanywhere, as far as I remember it was easy to setup[0].

[0] https://help.pythonanywhere.com/pages/FollowingTheDjangoTutorial

**Comment by Sanjay-B** [comment]
What exactly is that? Ive heard of it. So, to be clear, you cannot use a subdomain as an "ALLOWED HOSTS" or can you use the subdomain they gave me as an "ALLOWED HOSTS"?

**Comment by nitely** [comment]
> What exactly is that? Ive heard of it

A hosting provider.

> you cannot use a subdomain as an "ALLOWED HOSTS"

Yes, you can use a subdomain. You can use the subdomain a hosting provider gives you.

What I meant was that you can not use that subdomain for *other* web hostings. If you are hosting your site on heroku you won't be able to use `mysite.heliohost.org`. Heroku gives their own subdomains ie: `mysite.heroku.com` as does every other company offering a free plan (pythonanywhere, openshift, etc). Since you basically don't own those domains you can point them to other hosting providers. If you want your own domain you have to buy one.

**Comment by Sanjay-B** [comment]
Well, that's kinda sad. Im nowhere near dishing out money. :(

**Comment by nitely** [comment]
If you use pythonanywhere you get a subdomain `myforum.pythonanywhere.com` for free...

**Comment by Sanjay-B** [comment]
I might just host a web server with a spare computer I have. Then, I could connect it to the MySQL database I have. That should probably solve a majority of problems. 🤔

**Comment by nitely** [comment]
Only if you have fixed IP. If your IP changes everytime you restart your modem then you are out of luck. This is basically coz a domain needs to point to your IP and IP changes takes about 48hs to propagate over the world. You also need a domain, there a few free ones like http://www.noip.com/free

**Comment by Sanjay-B** [comment]
So im using BIND to host a DNS server, and yes, my IP is static. I had to recently do this for a game server I was using for a couple mates of mine. :)

**Comment by nitely** [comment]
Great. Then you should try noip.com if you want to browse your site through `myforum.no-ip.org` instead of typing the IP.

AFAIK you could use `myforum.no-ip.org` for heroku for example. Not sure about pythonanywhere, but probably the same as well.

**Comment by Sanjay-B** [comment]
Isnt ``myforum.no-ip.org`` still a sub-domain though? I'm talking about hosting a ``www.domainname.com`` or ``www.domainname.io`` .

**Comment by nitely** [comment]
It is, but at least you keep it if ever move to another hosting. `myforum.com`, `myforum.io`, etc are paid ones.

**Comment by Sanjay-B** [comment]
Very true. Thanks so much! :)

**Comment by Sanjay-B** [comment]
By chance, do you know how to run a django app in WAMP?

**Comment by nitely** [comment]
I've not used WAMP. But I know you need `mod_wsgi` to run python apps in Apache. Django has docs for it[0]. But IMHO the easiest way to run django is Nginx + gunicorn, there are many tutorial for this.

[0] https://docs.djangoproject.com/en/1.10/howto/deployment/wsgi/modwsgi/

**Comment by Sanjay-B** [comment]
What's the difference between the two?

**Comment by nitely** [comment]
Nginx is very lightweight and fast. It's specially good for serving static files (media files, images, etc) and it can be used as a reverse proxy to pass requests to gunicorn (an http server for python). WTServer seems to be the alternative to WAMP for nginx.

**Comment by nitely** [comment]
I would say use WAMP and if you find it slow or heavy on resources, then try WTServer.

**Comment by Sanjay-B** [comment]
The problem I see with WAMP is that there's no documentation or tutorials on how to use your own "paid" domain. Its more of the "localhost" type but I have seen it been able to do what I want it to. Its just a matter of how to do it.

**Comment by nitely** [comment]
Probably coz all web servers (hosting or local) run on localhost (127.0.0.1) and port 80 (the default for http) or 443 (the default for https). Then you gotta configure your Domain to point to your IP, usually from an admin panel provided by the company selling domains (ie namecheap.com or no-ip.com). Then you gotta configure bind9 as a DNS server.

**Comment by Sanjay-B** [comment]
Which IP? The 127.x.x or my public one?

**Comment by nitely** [comment]
The public one.

**Comment by nitely** [comment]
If you've a router, you gotta do port forwarding for *incoming* traffic on port 80 to the machine running the web server.

btw, you may not need bind9. I use it as reverse dns for SMTP (sending emails) and to serve host records, but domain providers like namecheap.com offer all that.

**Comment by Sanjay-B** [comment]
I got almost everything to work. I ended up using XAMPP and the port is 8080. To mask my IP, I used no-ip to create a free "domain". How exactly would I host the django app now? I am running my webserver locally and port forwarded 8080.

**Comment by nitely** [comment]
Check your app is working at 127.0.0.1:8080 (or localhost:8080), then check
you can access from the outside at http://publicIP (or publicIP:80 should
be the same), then configure no-ip and wait 48hs (sometimes it takes ~24hs)
and try the subdomain.

To run python/django on apache you just need the module mod_wsgi , read the
django docs I linked earlier. There are many tutorials though[0]

[0]
https://www.devbattles.com/en/sand/post-304-using+Apache+and+Python+WSGI+on+Windows

**Comment by Sanjay-B** [comment]
Whenever I activate that module, I get this error. 
``[Sat Mar 04 15:34:25.651884 2017] [authz_core:error] [pid 4080:tid 1864] [client 98.230.191.66:38927] AH01630: client denied by server configuration: C:/xampp/webalizer``
When deactivated, server works fine.

**Comment by nitely** [comment]
Did you edit `httpd.conf` and added the wsgi directoy, etc?

**Comment by nitely** [comment]
Try this as well: http://thelinuxfaq.com/453-apache-client-denied-by-server-configuration-in-xampp
Replacing by `Require all granted` should work...

**Comment by Sanjay-B** [comment]
```
9:14:23 AM  [Apache] 	Error: Apache shutdown unexpectedly.
9:14:23 AM  [Apache] 	This may be due to a blocked port, missing dependencies, 
9:14:23 AM  [Apache] 	improper privileges, a crash, or a shutdown by another method.
9:14:23 AM  [Apache] 	Press the Logs button to view error logs and check
9:14:23 AM  [Apache] 	the Windows Event Viewer for more clues
9:14:23 AM  [Apache] 	If you need more help, copy and post this
9:14:23 AM  [Apache] 	entire log window on the forum
```

**Comment by Sanjay-B** [comment]
Its the same error. I dont understand how nothing seems to be working. XAMPP no longer works for me. :(

**Comment by nitely** [comment]
No idea.

Apparently errors can be related to a variety of things... wsig module not matching the installed python version, [not matching the same VC version](http://stackoverflow.com/questions/28956767/apache-2-4-python-and-django-on-windows), [saving the httpd.conf file in a encoding other than utf-8](http://stackoverflow.com/questions/26721190/apache-shutdown-unexpectedly-after-documentroot-change-in-httpd-conf), etc. The usual for Windows users I guess, haha.

**Comment by nitely** [comment]
There should be an apache error log somewhere. I mean one that actually shows the real error.

**Comment by nitely** [comment]
> Press the Logs button to view error logs and check the Windows Event Viewer

Check those two places...

**Comment by nitely** [comment]
Here's another thing to check up: [Python must be the 32bit version and must be installed for all users](http://serverfault.com/questions/230635/wsgi-xampp-and-windows)

**Comment by nitely** [comment]
[Here is a video of someone setting up everything for an older version of XAMPP](https://www.youtube.com/watch?v=VOkXG0jL_8U) it may be useful. Apparently you also have to add python path to the system path (shown in the video).

Or you know... dual boot Ubuntu :wink:

**Comment by nitely** [comment]
> XAMPP no longer works for me. :(

That sounds like a syntax error in the conf file. If you happen to have non US-ascii chars in your directory path then it's very likely the problem. You should get a new copy. Edit it with Notepad++ and save it as utf-8 (it will probably open as utf-8, so just save it).

**Comment by Sanjay-B** [comment]
After 3 hours, still same error. This may be a bit much to ask but i'm out of options here. Could you please do it for me? Would help me loads if you could.

**Comment by nitely** [comment]
Well, I don't have much free time. If you can post the Apache logs here I may be of help. Although, I've a windows virtual machine laying around, so I could try XAMPP and see if I can get it working... I'll let you know.

**Comment by Sanjay-B** [comment]
Error log 1
```
6:16:20 PM  [Apache] 	Error: Apache shutdown unexpectedly.
6:16:20 PM  [Apache] 	This may be due to a blocked port, missing dependencies, 
6:16:20 PM  [Apache] 	improper privileges, a crash, or a shutdown by another method.
6:16:20 PM  [Apache] 	Press the Logs button to view error logs and check
6:16:20 PM  [Apache] 	the Windows Event Viewer for more clues
6:16:20 PM  [Apache] 	If you need more help, copy and post this
6:16:20 PM  [Apache] 	entire log window on the forums
```

Error Log 2 
```
[Sat Mar 04 15:34:25.651884 2017] [authz_core:error] [pid 4080:tid 1864] [client 98.230.191.66:38927] AH01630: client denied by server configuration: C:/xampp/webalize
```

**Comment by nitely** [comment]
What xampp version are you running ?

**Comment by Sanjay-B** [comment]
XAMPP 2.4. Thanks so much. :)

**Comment by nitely** [comment]
That's the apache version, no? Latest xampp is 7.1.1... There is also a
v5.6.3 in their site

**Comment by Sanjay-B** [comment]
Im stupid. Its 7.1.1. Lel

**Comment by nitely** [comment]
Got it working. I had all the same issue as you.

Here is what I did:

Download the `mod_wsgi` from http://www.lfd.uci.edu/~gohlke/pythonlibs/#mod_wsgi . It must match VC14. This one is for Python 3.6 (cp36) x86 (win32)
```
mod_wsgi-4.4.23+ap24vc14-cp36-cp36m-win32.whl
```
Extract it (ie with 7-zip, or rename the file to .zip and try winzip)
Copy/past the `mod_wsgi.so` into `apache/modules`.
Download python 3.6 x86, while installing make sure to check the "add to ENV path" in the first screen.

Add the following line into httpd.conf
```
LoadModule wsgi_module modules/mod_wsgi.so
```

All done.

I found I had a bunch of duplicated xampp process running and kill them all. Probably due to stuff I was trying. Better to just restart the PC.

If there is any error, then try running `xampp/xampp_start.exe` , the error.log and event viewer are useless.
